### DISCUSSION TITLE
Server log analysis

### DISCUSSION THREAD

**Discussion by simsong** [plausible/analytics#716] [Anything Else]
Can I load log files from Apache or Amazon S3? A lot of my pages are not loaded via HTML with embedded JavaScript.

For example, can I write a program to scan my logfiles and incorporate the data directly into the RDBMS?

**Comment by metmarkosaric** [comment]
that's not possible @simsong. server logs are very inaccurate as they include mostly spam and bots so the data we can get from them is not that useful. you can read more here: https://plausible.io/blog/server-log-analysis

**Reply by simsong** [reply]
It depends on your purpose. From my point of view server logs are very accurate because they contain records of bots and automated downloaders, which are of great interest to me in my application. I wish to distinguish the behavior of humans and automated crawlers on my site, which is used by both.

**Reply by metmarkosaric** [reply]
ahh I see. that's a bit of a different use case compared to what we normally do with Plausible where we try to count humans only and exclude any bots.

**Reply by simsong** [reply]
> ahh I see. that's a bit of a different use case compared to what we normally do with Plausible where we try to count humans only and exclude any bots.

Bots are just programs written by humans performing a function for the humans. I run a digital forensics website (https://digitalcorpora.org) and it's quite common for people (students) to want to download large numbers of files. They write programs to do this.

**Comment by ukutaht** [comment]
Supporting server logs would be great but it's a lot more difficult to make it work than JS-based analytics. We don't have hard plans right now but something to consider in the future

**Reply by simsong** [reply]
What precisely is the difficulty? On Apache I can put together sessions from the same browser with very high success.

**Reply by metmarkosaric** [reply]
Bot detection. To filter out bots and spam would be really difficult compared to the JS-based stats. I shared a blog post URL which has full details on how large problem this is for server logs.

**Reply by simsong** [reply]
But I don't want to filter out bots. I want to know about them.

**Reply by simsong** [reply]
Also, I how hard can it be to write a bot that gets by your JavaScript-based bot-detection? Just run it in chromium and use a decent markov model to simulate mouse movement and clicking.

**Reply by metmarkosaric** [reply]
i understand. it's a use case we can consider in the future. 

for now, we do a lot to exclude bots as that's what the normal use case is. obviously you can always do something to avoid the bot detection but that is not very common so not a big issue. including all the normal bots that visit every website would ruin the dashboard for our normal users who only want to check human visitors.

**Reply by simsong** [reply]
It doesn’t need to ruin the dashboard; you just need a tab to show bots and non-bots. For a website concerned about its bandwidth, which is an issue for many outside of the developed world, or for those who self-host, or host out of s3, knowing how the bandwidth balances between bots and non-bots is very important.

**Comment by 1000i100** [comment]
Like @simsong i'm interested in server-log based analytics.

My politic on the (static generated) website I build is : 

- Zero cookies
- All content must be available without JS. JS can add some comfort feature, but must be truly optional.
- Zero query outside the website domain

And my clients love the speed, simplicity and robustness it gives to my websites.

I would like to track : 

- usual human traffic (position based on IP, if there are on mobile or computer with userAgent, and their path on the site when they go in. I'm not very interested to know if it's the same people a day and another on the site)
- blind people trying to browse my website with voice reader or braille devices
- seo-bot traffic (to better understand bot behaviors and consequence in SEO : why some resources are indexed and updated, and why other not, and track associated bandwidth)
- hostile-bot (spam-bot, scan for vulnerability bot... and associated bandwidth and IP-range to build dynamic block-list)
- RSS/atom usage
- and when I do other thing than static websites, I would like to use the same UI to track my API usages (REST, GraphQL, Elasticsearch)

For all this, a documented at least basic server-log-analysis way to feed Plausible would be a must !

**Reply by simsong** [reply]
Wow, I had completely forgotten about this. Thanks for tagging me. I've pretty much decided to build analytics into my back-end.

**Reply by 1000i100** [reply]
I will perhaps do the same if I not found what I search.
In Open Source self-hostable server log centered, I've found :
- https://goaccess.io/
- https://www.awstats.org/
- https://posthog.com/
- https://grafana.com/
- https://prometheus.io/

Not centered on it but able to do it, there is : https://matomo.org/  and  https://count.ly/
Perhapse also : http://www.openwebanalytics.com/ but I've not tested it.

And more monitoring oriented : 
- https://monika.hyperjump.tech/

And some real time fun viewer with other use case than mine : 
- http://www.fudgie.org/gltrail.html
- https://logstalgia.io/

Most of them found via : 
https://alternativeto.net/category/networking-and-admin/web-log-analyzer/?license=opensource
and i've also checked here : https://plausible.io/blog/google-analytics-alternatives#open-source-and-self-hosted-google-analytics-alternatives

**Comment by srsree** [comment]
Please consider/ include the server log analysis feature. Many of our requests are served as APIs, and we will have to have lerver log analysis available to process them.
